### ISSUE TITLE
[BUG] Performance regression after the refactory that moves losses and metrics to model.compile()

### ISSUE THREAD

**Issue by gabrielspmoreira** [NVIDIA-Merlin/models#491]
### Bug description
We observed a strong performance (throughput) regression after the refactory that moved losses and metrics to `model.compile()`.
The *average examples / sec* metric dropped from ~105k to 42k. That can be reproduced using the [retrieval experiments scripts](https://github.com/NVIDIA-Merlin/research/blob/main/retrieval_exp/scripts/retrieval_train_eval.py) with the same arguments, only moving the loss and metrics out of the task to the `model.compile()`.

After using some commits to isolate the issue as suggested by @marcromeyn , it seems the problem is introduced by some commit after this [commit [A]](https://github.com/NVIDIA-Merlin/models/commit/168ad03f8089146647c23b5fbe5fc65b39779ecb) and before (including) this [commit [B]](https://github.com/NVIDIA-Merlin/models/commit/10a3b8132c89991e7d25b0101c8e811654f24458).

Isolating a bit the issue, the lower throughput seems to be related to the metrics computation during training, as I also tested the pipeline (1) after removing all ranking metrics eval during training (Recall, NDCG, MRR) and (2) removing all metrics, including custom ones where I aggregate statistics from the logits.

In the following plot, it is possible to see the throughput at commit A (before the change) in **brown**, the throughput at commit B in **purple**, the throughput after removing ranking metrics (1) in **blue**, and the throughput after removing all but one metric (2) in **green**, which is closer to the original throughput (brown) 

![image](https://user-images.githubusercontent.com/504752/171931543-dd49b5c8-18c4-40de-967a-a7841741a972.png)

So I suspect that some mechanisms might not be working properly after the refactory:
1 - The `train_metrics_steps` arg for `model.fit()` that configures training to compute metric values only each n steps is not working properly after the refactory.
2 Maybe that mechanism that sort predictions and targets only once for all ranking metrics and different cut-offs is not working anymore

More details on these perf tests in this [Slack thread (Nvidia Merlin internal)](https://nvidia.slack.com/archives/C020P24SQJJ/p1654152215076019)
